### PR TITLE
Add fp16 backward support

### PR DESCRIPTION
**Pull request by chengduoZH**
**This PR's work includes:**
- Make `sum`, `cross_entropy`, `mean`, `scale`, `cross_entropy_grad`, `softmax_grad`, `mean_grad`, `mul_grad `, `pool_cudnn_grad`, `batch_norm_cudnn_grad`, `conv_cudnn_grad` , `elementwise_add_grad` support FP16.
- Add unit test for `sum_op`, `softmax_op`, `scale_op`, `mean_op`, `mul_op`, `cross_entropy_op` for FP16.
- Add fp16 unit test for sum_op, activation_op, scale_op, pool_cudnn_op, cross_entropy_op, mul_op, softmax_grad_op.

### CODE REVIEW

File: python/paddle/fluid/tests/unittests/test_conv2d_op.py [hunk 1]
```diff
@@ -223,106 +223,81 @@ def init_group(self):
...
+create_test_cudnn_fp16_class(
+    TestWithPad, "TestPool2DCUDNNFp16OpCase1", grad_check=False)
+create_test_cudnn_fp16_class(
+    TestWithStride, "TestPool2DCUDNNFp16OpCase2", grad_check=False)
+create_test_cudnn_fp16_class(
+    TestWithGroup, "TestPool2DCUDNNFp16OpCase3", grad_check=False)
+create_test_cudnn_fp16_class(
+    TestWith1x1, "TestPool2DCUDNNFp16OpCase4", grad_check=False)
+create_test_cudnn_fp16_class(
+    TestWithInput1x1Filter1x1, "TestPool2DCUDNNFp16OpCase4", grad_check=False)
```
Comment (chengduoZH) [inline python/paddle/fluid/tests/unittests/test_conv2d_op.py:298]: Grad check fails for conv_cudnn, the `max_relative_error ` is [0.095, 0.992].

File: paddle/fluid/operators/math/cross_entropy.h [hunk 2]
```diff
@@ -33,6 +34,19 @@ struct TolerableValue {
   }
 };
 
+// float16 value clip behave different.
```
Comment (dzhwinter) [inline paddle/fluid/operators/math/cross_entropy.h]: 补充注释了，重新copy一下吧

File: paddle/fluid/operators/mean_op.h [hunk 3]
```diff
@@ -55,8 +55,8 @@ class MeanGradKernel : public framework::OpKernel<T> {
     IG->mutable_data<T>(context.GetPlace());
 
     T ig_size = static_cast<T>(IG->numel());
-    Eigen::DSizes<int, 1> bcast(ig_size);
-
+    Eigen::DSizes<int, 1> bcast(static_cast<int>(ig_size));
+    // EigenVector<T>::From can cast float to float16?
```
Comment (dzhwinter) [inline paddle/fluid/operators/mean_op.h]: seems no cast needed?

File: python/paddle/fluid/tests/unittests/test_activation_op.py [hunk 4]
```diff
@@ -1067,28 +578,70 @@ def setUp(self):
...
         self.check_grad(['X'], 'Out', max_relative_error=0.008)
 
-    def init_dtype(self):
-        pass
-
 
-class TestFP16Swish(TestSwish):
-    def init_dtype(self):
-        self.dtype = np.float16
+#------------------ Test Fp16 ----------------------
```
Comment (dzhwinter) [inline python/paddle/fluid/tests/unittests/test_activation_op.py:587]: why merge the float16 tests into one tests?  It is not a standalone tests, and may lead to a longer test runtime if the float16 op increasing.
Just modify the op_test base class, and leave the float16 tests individual.

**Review by dzhwinter — DISMISSED** [review]
LGTM. leave some comments.

File: python/paddle/fluid/tests/unittests/test_activation_op.py (hunk 4 above)
Comment (chengduoZH) [reply python/paddle/fluid/tests/unittests/test_activation_op.py:587]: I have compared the time consuming between the current mode and before, they are nearly equal, about `9.85 sec`. And after adding gradient check for fp16, test_activation_op need `11.47 sec`

File: paddle/fluid/operators/mean_op.h (hunk 3 above)
Comment (chengduoZH) [reply paddle/fluid/operators/mean_op.h]: It is necessary, otherwise the compile throw errors.

File: paddle/fluid/operators/math/cross_entropy.h (hunk 2 above)
Comment (chengduoZH) [reply paddle/fluid/operators/math/cross_entropy.h]: Done

**Review by dzhwinter — APPROVED** [review]
Good Job.
